pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,410 @@
+	if err != nil {
+		return false, nil, -1, err
+	}
+	if !present {
+		rc, size, err := s.source.GetBlob(ctx, blobinfo)
+		if err != nil {
+			return false, nil, -1, errors.Wrapf(err, "error reading blob from source image %q", transports.ImageName(s.reference))
+		}
+		return false, rc, size, nil
+	}
+	for _, directory := range s.reference.directories {
+		if directory == "" {
+			continue
+		}
+		for _, isConfig := range []bool{false, true} {
+			filename := filepath.Join(directory, makeFilename(blobinfo.Digest, isConfig))
+			f, err := os.Open(filename)
+			if err == nil {
+				return true, f, size, nil
+			}
+		}
+	}
+	rc, size, err := s.source.GetBlob(ctx, blobinfo)
+	if err != nil {
+		return false, rc, size, errors.Wrapf(err, "error reading blob from source image %q", transports.ImageName(s.reference))
+	}
+	return false, rc, size, nil
+}
+
+func (s *blobCacheSource) GetBlob(ctx context.Context, blobinfo types.BlobInfo) (io.ReadCloser, int64, error) {
+	_, rc, size, err := s.getBlob(ctx, blobinfo)
+	return rc, size, err
+}
+
+func (s *blobCacheSource) GetSignatures(ctx context.Context, instanceDigest *digest.Digest) ([][]byte, error) {
+	return s.source.GetSignatures(ctx, instanceDigest)
+}
+
+func (s *blobCacheSource) LayerInfosForCopy(ctx context.Context) ([]types.BlobInfo, error) {
+	return nil, nil
+}
+
+func (d *blobCacheDestination) Reference() types.ImageReference {
+	return d.reference
+}
+
+func (d *blobCacheDestination) Close() error {
+	return d.destination.Close()
+}
+
+func (d *blobCacheDestination) SupportedManifestMIMETypes() []string {
+	return d.destination.SupportedManifestMIMETypes()
+}
+
+func (d *blobCacheDestination) SupportsSignatures(ctx context.Context) error {
+	return d.destination.SupportsSignatures(ctx)
+}
+
+func (d *blobCacheDestination) DesiredLayerCompression() types.LayerCompression {
+	return d.destination.DesiredLayerCompression()
+}
+
+func (d *blobCacheDestination) AcceptsForeignLayerURLs() bool {
+	return d.destination.AcceptsForeignLayerURLs()
+}
+
+func (d *blobCacheDestination) MustMatchRuntimeOS() bool {
+	return d.destination.MustMatchRuntimeOS()
+}
+
+func (d *blobCacheDestination) IgnoresEmbeddedDockerReference() bool {
+	return d.destination.IgnoresEmbeddedDockerReference()
+}
+
+func (d *blobCacheDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, isConfig bool) (types.BlobInfo, error) {
+	if inputInfo.Digest != "" && len(d.reference.directories) > 0 {
+		if d.reference.removeFirstDirectoryOnCleanup && d.reference.directories[0] == "" {
+			temp, err := ioutil.TempDir("", "")
+			if err != nil {
+				return types.BlobInfo{}, errors.Wrapf(err, "error creating new temporary directory for blob cache")
+			}
+			d.reference.directories[0] = temp
+		}
+		filename := filepath.Join(d.reference.directories[0], makeFilename(inputInfo.Digest, isConfig))
+		f, err := os.OpenFile(filename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+		if err == nil {
```

Comment:
Are you ignoring err != nil?
Should we only ignore ENOEXIST?

Reply:
Yes, and probably yes.  Adding a debug message for the `!os.IsNotExist()` case.